### PR TITLE
RBF: First part

### PR DESCRIPTION
**Pull request by niftynei**
These patches have been indescriminately split into 3 parts to make it a bit easier(?) to review.

This section adds the new `inflight` idea and starts moving away from `uncommitted_channel`.

### CODE REVIEW

File: lightningd/dual_open_control.c [hunk 1]
```diff
@@ -1601,6 +1601,85 @@ static void handle_peer_tx_sigs_msg(struct subd *dualopend,
...
+
+	if (!fromwire_dualopend_rbf_validate(tmpctx, msg,
+					     &candidate_psbt)) {
+		channel_internal_error(channel,
+				       "Malformed dualopend_rbf_validate %s",
+				       tal_hex(tmpctx, msg));
+		return;
+	}
+
+	/* Iterate through all previous inflights, confirm
```
Comment (rustyrussell) [inline lightningd/dual_open_control.c]: IMHO This logic is cleaner if we start with the new candidate, then iterate.  And we only need a flag per input, not an actual PSBT.

Approx:
```
bool *inputs_present = tal_arr(tmpctx, set_psbt->num_inputs);
memset(inputs_present, true, tal_bytelen(inputs_present));

list_for_each(&channel->inflights, inflight, list) {
    for (size_t i = 0; i < set_psbt->num_inputs; i++) {
       if (!psbt_has_input(inflight->funding_psbt, set_psbt->tx->inputs[i]))
            inputs_present[i] = false;
    }
}

if (memeqzero(inputs_present, tal_bytelen(inputs_present)) {
   error ...
```

**Review by rustyrussell — APPROVED** [review]
Minor things to fix up, but looking good so far!

File: openingd/dualopend.c [hunk 2]
```diff
@@ -2349,6 +2346,180 @@ static void opener_start(struct state *state, u8 *msg)
...
+	 * 512 + 512 / 4 or 640 sat/kw, `fee_step` 2
+	 * is 640 + 640 / 4 or 800 sat/kw.
+	 */
+	for (; fee_step > 0; fee_step--)
+		feerate += feerate / 4;
+
+	/* It's possible they sent us a 'bad' feerate step,
+	 * i.e. less than the last one. */
+	if (feerate < last_feerate)
+		return false;
```
[the remark this answers is not in the document]
Comment (niftynei) [reply openingd/dualopend.c:2419]: Do we care if they're retrying the same amount? It'll (most likely) fail when we check that the net fees is greater than the last.

File: lightningd/channel.c [hunk 3]
```diff
@@ -143,6 +144,45 @@ void get_channel_basepoints(struct lightningd *ld,
...
+	struct channel_inflight inf_init = {
+		.channel = channel,
+		.funding_txid = funding_txid,
+		.funding = funding,
+		.funding_outnum = funding_outnum,
+		.funding_feerate = funding_feerate,
+		.our_funds = our_funds,
+		.remote_tx_sigs = false,
+	};
+	memcpy(inflight, &inf_init, sizeof(struct channel_inflight));
```
[the remark this answers is not in the document]
Comment (niftynei) [reply lightningd/channel.c:176]: > ```
> inflight->channel = channel;
> inflight->funding_txid = funding_txid;
> ...

These fields are marked `const` on the struct declaration, so this method fails to compile. Is there a way around the compilation error?


```
lightningd/channel.c: In function ‘new_inflight’:
lightningd/channel.c:168:25: error: assignment of read-only member ‘funding_txid’
  168 |  inflight->funding_txid = funding_txid;
      |                         ^
lightningd/channel.c:169:20: error: assignment of read-only member ‘funding’
  169 |  inflight->funding = funding;
      |                    ^
lightningd/channel.c:170:27: error: assignment of read-only member ‘funding_outnum’
  170 |  inflight->funding_outnum = funding_outnum;
      |                           ^
lightningd/channel.c:171:28: error: assignment of read-only member ‘funding_feerate’
  171 |  inflight->funding_feerate = funding_feerate;
      |                            ^
lightningd/channel.c:172:22: error: assignment of read-only member ‘our_funds’
  172 |  inflight->our_funds = our_funds;
      |                      ^

```

File: openingd/dualopend.c [hunk 4]
```diff
@@ -2349,6 +2346,180 @@ static void opener_start(struct state *state, u8 *msg)
...
+	/* We need a new tx_state! */
+	tx_state = new_tx_state(state);
+
+	if (!fromwire_init_rbf(rbf_msg, &cid,
+			       &tx_state->opener_funding,
+			       &tx_state->tx_locktime,
+			       &fee_step))
+		peer_failed_err(state->pps, &state->channel_id,
+				"Parsing init_rbf %s",
+				tal_hex(tmpctx, rbf_msg));
```
[the remark this answers is not in the document]
Comment (niftynei) [reply openingd/dualopend.c:2445]: I chose to "hard fail" for any message parsing failures, with the idea that if your peer is sending fubar'd message data formats might as well close the channel.

File: lightningd/dual_open_control.c (hunk 1 above)
Comment (niftynei) [reply lightningd/dual_open_control.c]: This is much nicer!

**Comment by niftynei** [comment]
Everything but the memcpy/const comment has been addressed.

Note that now includes patch from #4402, to get rid of weird hang when skipping one of the v2 tests.

File: lightningd/channel.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/channel.c:176]: Ah, right!  I would usually create a struct funding_tx, and use a const pointer here.  That gives a more natural method than construct & memcpy, though technically it's less efficient.

File: openingd/dualopend.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (rustyrussell) [reply openingd/dualopend.c:2429]: The purpose of defined steps was to stop them spamming, so seems simpler to insist on an increase?

File: openingd/dualopend.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (rustyrussell) [reply openingd/dualopend.c:2455]: Yeah, I've been tending to "don't fail channel unless we *have* to" with the latest changes, however.  Ofc it should Never Happen...

File: lightningd/dual_open_control.c [hunk 5]
```diff
@@ -1568,8 +1597,118 @@ static void handle_peer_tx_sigs_msg(struct subd *dualopend,
...
+					     &candidate_psbt)) {
+		channel_internal_error(channel,
+				       "Malformed dualopend_rbf_validate %s",
+				       tal_hex(tmpctx, msg));
+		return;
+	}
+
+	/* Grab the first ever funding tx as starting point */
+	inflight = list_top(&channel->inflights, struct channel_inflight, list);
+	set_psbt = clone_psbt(NULL, inflight->funding_psbt);
```
Comment (rustyrussell) [inline lightningd/dual_open_control.c:1623]: Don't need this clone any more, I think?

**Review by rustyrussell — APPROVED** [review]
All good except for one change (remove unnecessary clone).

File: openingd/dualopend.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (niftynei) [reply openingd/dualopend.c:2429]: Sure, that's easy enough to enforce!

File: openingd/dualopend.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (niftynei) [reply openingd/dualopend.c:2455]: Right! I was counting "bad message encoding" as a *have to* haha

File: lightningd/dual_open_control.c (hunk 5 above)
Comment (niftynei) [reply lightningd/dual_open_control.c:1623]: we need a copy of the original PSBT to run comparisons against; we re-use `inflight` to iterate below.

Comment (rustyrussell) [reply lightningd/dual_open_control.c:1623]: I don't think so.  Start with the candidate, compare it to the inflights:

```diff
diff --git a/lightningd/dual_open_control.c b/lightningd/dual_open_control.c
index 89a17ff63..fc8b01e55 100644
--- a/lightningd/dual_open_control.c
+++ b/lightningd/dual_open_control.c
@@ -1606,7 +1606,6 @@ static void handle_validate_rbf(struct subd *dualopend,
 	struct wally_psbt *candidate_psbt;
 	struct channel_inflight *inflight;
 	struct channel *channel = dualopend->channel;
-	struct wally_psbt *set_psbt;
 	bool *inputs_present;
 	struct amount_sat candidate_fee, last_fee;
 
@@ -1618,11 +1617,7 @@ static void handle_validate_rbf(struct subd *dualopend,
 		return;
 	}
 
-	/* Grab the first ever funding tx as starting point */
-	inflight = list_top(&channel->inflights, struct channel_inflight, list);
-	set_psbt = clone_psbt(NULL, inflight->funding_psbt);
-
-	inputs_present = tal_arr(tmpctx, bool, set_psbt->num_inputs);
+	inputs_present = tal_arr(tmpctx, bool, candidate_psbt->num_inputs);
 	memset(inputs_present, true, tal_bytelen(inputs_present));
 
 	/* Iterate through all previous inflights, confirm
@@ -1630,9 +1625,9 @@ static void handle_validate_rbf(struct subd *dualopend,
 	 * this one. */
 	list_for_each(&channel->inflights, inflight, list) {
 		/* Remove every non-matching input from set */
-		for (size_t i = 0; i < set_psbt->num_inputs; i++) {
+		for (size_t i = 0; i < candidate_psbt->num_inputs; i++) {
 			struct wally_tx_input *input =
-				&set_psbt->tx->inputs[i];
+				&candidate_psbt->tx->inputs[i];
 			struct bitcoin_txid in_txid;
 
 			wally_tx_input_get_txid(input, &in_txid);
@@ -1643,17 +1638,6 @@ static void handle_validate_rbf(struct subd *dualopend,
 		}
 	}
 
-	/* Finally, let's check the candidate psbt */
-	for (size_t i = 0; i < set_psbt->num_inputs; i++) {
-		struct wally_tx_input *input = &set_psbt->tx->inputs[i];
-		struct bitcoin_txid in_txid;
-
-		wally_tx_input_get_txid(input, &in_txid);
-
-		if (!psbt_has_input(candidate_psbt, &in_txid, input->index))
-			inputs_present[i] = false;
-	}
-
 	/* Are there any inputs that were present on all txs? */
 	if (memeqzero(inputs_present, tal_bytelen(inputs_present))) {
 		char *errmsg;
@@ -1708,7 +1692,6 @@ static void handle_validate_rbf(struct subd *dualopend,
 
 send_msg:
 	subd_send_msg(channel->owner, take(msg));
-	tal_free(set_psbt);
 }
```

**Comment by rustyrussell** [comment]
Trivial rebase for gen file conflict.

**Comment by rustyrussell** [comment]
Argh, then I merged some others and clashed again...

**Pull request merged**